### PR TITLE
[Do Not Merge [yet]] Update jupyterlab

### LINKED ISSUE #65625: Jupyterlab 1.0.4 fails to build, citing missing `tornado` version which exists
**Describe the bug**

When trying to install the latest version of jupyterlab, nix fails, complaining about a missing library `tornado`, so long as it isn't 6.0 or above. Yet `tornado` installs just fine, and with a version of 5.X. Here's the error: 

```
/build/jupyterlab-1.0.4/dist /build/jupyterlab-1.0.4
Processing ./jupyterlab-1.0.4-py2.py3-none-any.whl
Collecting tornado!=6.0.0,!=6.0.1,!=6.0.2 (from jupyterlab==1.0.4)
  Could not find a version that satisfies the requirement tornado!=6.0.0,!=6.0.1,!=6.0.2 (from jupyterlab==1.0.4) (from versions: )
No matching distribution found for tornado!=6.0.0,!=6.0.1,!=6.0.2 (from jupyterlab==1.0.4)
builder for '/nix/store/bigclz0i9ld7dxj33vw1c8kyiwdrimjd-python3.7-jupyterlab-1.0.4.drv' failed with exit code 1
cannot build derivation '/nix/store/9vk437k3q0w79r1f6phxwhcap083j813-python3-3.7.3-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/9vk437k3q0w79r1f6phxwhcap083j813-python3-3.7.3-env.drv' failed

```

**To Reproduce**

1. Try to run `nix-shell` with this file: 

```
with import <nixpkgs> {};

( let
    my_jupyterlab = python37.pkgs.buildPythonPackage rec {
      pname = "jupyterlab";
      version = "1.0.4";

      src = python37.pkgs.fetchPypi {
        inherit pname version;
        sha256 = "0vhdbzrphirl3x37d4zh0gaxhs6f7wj0i30wy4878yjkmlr219rj";
      };

      doCheck = false;
    };

  in python37.withPackages (ps: [ps.jupyter ps.tornado my_jupyterlab])
).env

```
2. Note that it fails with the error above. 

**Expected behavior**
It would finish building the environment with Jupyterlab 1.0.4

**Additional context**

**Metadata**
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.
```
[1 copied, 0.0 MiB DL]
 - system: `"x86_64-linux"`
 - host os: `Linux 5.2.4, NixOS, 19.09pre186857.239fffc90d7 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(jon): `"home-manager, nixpkgs"`
 - channels(root): `"nixos-19.09pre186857.239fffc90d7"`
 - nixpkgs: `/home/jon/.nix-defexpr/channels/nixpkgs`

```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
 - python37Packages.jupyterlab
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by infinisil**
##### Motivation for this change

Do not merge yet, this is dependent on https://github.com/NixOS/nixpkgs/pull/67422

Closes https://github.com/NixOS/nixpkgs/issues/65625

This introduces `jsonschema3` in parallel to `jsonschema` (version 2), because apparently [a lot of python packages probably still depend on v2](https://github.com/NixOS/nixpkgs/pull/65824#issuecomment-517885153)

Ping @FRidh @JonathanReeve 

###### Things done

- [ ] Tested that the new version actually works. I'd like @JonathanReeve to do that because I'm not a jupyterlab user
- [x] Built using sandboxing
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by FRidh** [comment]
This jsonschema bump came by several times; we can only have a single
version.

**Comment by Infinisil** [comment]
@FRidh Why that? We need v3 for jupyterlab which is pretty popular. Would you prefer for v2 to be removed and replaced by v3?

**Comment by mmahut** [comment]
@GrahamcOfBorg build python37Packages.jupyterlab

**Comment by Infinisil** [comment]
Ah whoops, I only built jupyterlab_server previously, now jupyterlab itself also works

**Comment by mmahut** [comment]
@GrahamcOfBorg build python37Packages.jupyterlab

**Comment by rehno-lindeque** [comment]
I seem to be having some initial success trying [this approach](https://discourse.nixos.org/t/allowing-multiple-versions-of-python-package-in-pythonpath-nixpkgs/3849) suggested by @costrouc.

Quick/rough experiment looked like this:

```nix
let
  packageOverrides = selfPythonPackages: pythonPackages: {
    # v1.3.0 not yet in nixpkgs
    nixpkgs-pytools =
      selfPythonPackages.callPackage (
        { lib
        , buildPythonPackage
        , fetchPypi
        , jinja2
        , setuptools
        , isPy27
        }:

        buildPythonPackage rec {
          pname = "nixpkgs-pytools";
          version = "1.3.0";
          disabled = isPy27;

          src = fetchPypi {
            inherit pname version;
            sha256 = "11skcbi1lf9qcv9j5ikifb4pakhbbygqpcmv3390j7gxsa85cn19";
          };

          propagatedBuildInputs = [
            jinja2
            setuptools
            selfPythonPackages.rope
          ];

          # tests require network ..
          doCheck = false;
        }) {};

    jsonschema_3_0_2 = pythonPackages.buildPythonPackage rec {
      pname = "jsonschema";
      version = "3.0.2";

      src = pythonPackages.fetchPypi {
        inherit pname version;
        sha256 = "03anb4ljl624lixrsaxfi7i1iwavw39sd8cfkhcy0dr2dixjnjld";
      };

      nativeBuildInputs = with selfPythonPackages; [ setuptools_scm nixpkgs-pytools ];
      checkInputs = with selfPythonPackages; [ nose mock vcversioner twisted perf ];
      propagatedBuildInputs = with selfPythonPackages; [ functools32 attrs pyrsistent ];

      # Rewrite imports and alter setup.cfg
      postPatch = with selfPythonPackages; ''
        substituteInPlace jsonschema/tests/test_jsonschema_test_suite.py \
          --replace "python" "${pkgs.python3.pythonForBuild.interpreter}"

        substituteInPlace setup.cfg \
          --replace 'name = jsonschema' 'name = jsonschema_${src.outputHash}' \
          --replace 'jsonschema = schemas/*.json' 'jsonschema_${src.outputHash} = schemas/*.json'

        # Unsure if these are necessary
        # --replace 'jsonschema = jsonschema.cli:main' 'jsonschema_${src.outputHash} = jsonschema_${src.outputHash}.cli:main'
        # --replace 'jsonschema/__init__.py' 'jsonschema_${src.outputHash}/__init__.py'
        # --replace 'jsonschema/_reflect.py' 'jsonschema_${src.outputHash}/_reflect.py'

        substituteInPlace jsonschema/_utils.py \
          --replace 'pkgutil.get_data("jsonschema",' 'pkgutil.get_data("jsonschema_${jsonschema_3_0_2.src.outputHash}",'

        python-rewrite-imports \
          --path . \
          --replace jsonschema jsonschema_${jsonschema_3_0_2.src.outputHash}

      '';

      checkPhase = ''
        nosetests
      '';

      doCheck = false;
    };

    json5 = pythonPackages.buildPythonPackage rec {
      pname = "json5";
      version = "0.8.5";
      src = pythonPackages.fetchPypi {
        inherit pname version;
        sha256 = "1c3k5blbhq7g2lnbap26a846ag5x19ivisd3wfzz6bzdl46hyjqj";
      };
      doCheck = false;
    };

    jupyterlab_server = pythonPackages.buildPythonPackage rec {
      pname = "jupyterlab_server";
      version = "1.0.6";
      src = pythonPackages.fetchPypi {
        inherit pname version;
        sha256 = "1bax8iqwcc5p02h5ysdc48zvx7ll5jfzfsybhb3lfvyfpwkpb5yh";
      };
      nativeBuildInputs = with selfPythonPackages; [ nixpkgs-pytools ];
      propagatedBuildInputs = with selfPythonPackages; [
        notebook
        json5
        jsonschema_3_0_2
      ];
      doCheck = false;

      # Substitute the requirement for jsonschema in setup.py  and rewrite all imports.
      postPatch = with selfPythonPackages; ''
        substituteInPlace setup.py --replace "jsonschema>=3.0.1" "jsonschema_${jsonschema_3_0_2.src.outputHash}>=${jsonschema_3_0_2.version}"
        python-rewrite-imports \
          --path . \
          --replace jsonschema jsonschema_${jsonschema_3_0_2.src.outputHash}
      '';

    };
```

**Review by FRidh — CHANGES_REQUESTED** [review]
As we've discussed, no overrideScope' within the package set.

**Comment by Infinisil** [comment]
@FRidh I've updated this PR to put `jupyterlab` and `jupyterlab_server` in the top-level `pkgs.*` set, while replacing the ones in `pythonPackages.*` with a `throw` mentioning they've been moved.

I opted to still use `overrideScope'` however, because it's just less boilerplate. This is what I'm using to override the package set:
```nix
python3.pkgs.overrideScope' (self: super: {
  jsonschema = self.jsonschema3;
})
```

vs. what you'd need to write to achieve the same with `.override`:
```nix
(python3.override (old: {
  packageOverrides = lib.composeExtensions
    (old.packageOverrides or (self: super: {}))
    (self: super: {
      jsonschema = self.jsonschema3;
    });
})).pkgs
```

(if I leave out the `composeExtension` part, the previous `packageOverrides` of `python3` would be ignored)

So I think it's a good idea to merge https://github.com/NixOS/nixpkgs/pull/67422 because it simplifies this kind of thing a lot.

**Comment by costrouc** [comment]
@rheno-lindeque and @Infinisil the import rewrite method that I've demonstrated will work without overrides. I think it is worth thinking what will happen if we allow incompatible versions of packages into nixpkgs. I could provide a PR that demonstrates how this could be done without overrides?

**Comment by Infinisil** [comment]
@costrouc I'm not sure, but it seems to me that jupyterlab isn't meant to be used as a python package at all. So this PR now just moves it out of the python package set to `pkgs.jupyterlab`. I guess there's still the problem of the implicit python path thing when you set python packages in `buildInputs` (this is what you mean right?), but I think the fact that `buildInputs` can be used for python packages is bad design anyways, and I'd prefer people use `pythonPath = ...` or `python.withPackages` instead, then the `PYTHONPATH` problem shouldn't occur either (because you'd use `buildInputs = [ jupyterlab (python.withPackages ...) ]` which shouldn't conflate different versions).

I'm no python expert though, let me know if I got something mixed up. And feel free to demonstrate your approach if you feel like it fits in nixpkgs, I'm not quite sure myself what this entails.

**Comment by rehno-lindeque** [comment]
I think the link I posted may have gotten lost in the noise. @costrouc posted a detailed description over here:

* https://discourse.nixos.org/t/allowing-multiple-versions-of-python-package-in-pythonpath-nixpkgs/3849 
* An example here: https://github.com/costrouc/python-multiple-versions
* The code I posted above uses `python-rewrite-imports` which is the main technique he proposes.

As a proof point, I did manage to get jupyterlab 1.1.1 working using `python-rewrite-imports`, though I think it's a little unfortunate you have to use it in both the package as well as packages that depend on it to make this work... (I ended up abandoning my attempt because of difficulties with extensions in jupyterlab > 1.0.0, it worked otherwise)

I suppose a detailed discussion probably wants to go to discourse or an rfc though. I feel as though both approaches have merit, but I'm not a regular contributor.

**Comment by cdyson37** [comment]
Possibly stupid question: with jupyterlab as a standalone package, how does it "see" other Python modules? With `python37.withPackages (p: with p; [jupyterlab fancy_module])` I'd expect `fancy_module` to be available within the jupyterlab session. But if it's standalone presumably I need to tell it where to find a suitable kernel?

FWIW jsonschema is only really used in one place within jupyterlab_server as far as I can see, and is probably easily manually renamed with a bit of sed-magic; that might be the path of least resistance?

**Comment by MMesch** [comment]
@Infinisil , thank you very much for this helpful PR. Do you still work on it? Is it blocked on something specific?

**Comment by Infinisil** [comment]
For one, this PR still depends on #67442, which @FRidh wants docs for, but I really hate writing xml. I could change this PR to not depend on it though by rewriting it in the verbose style described in https://github.com/NixOS/nixpkgs/pull/67423#issuecomment-528137652.

Also I'm not using jupyterlab, and https://github.com/NixOS/nixpkgs/pull/67423#issuecomment-538420909 mentions that apparently you need to be able to add python modules inside it. I think this PR might break the workflow described there because it can lead to package conflicts for jsonschema. Maybe that's unproblematic though.

**Comment by tbenst** [comment]
as a heads up, jupyterlab is currently broken on master `nix-shell -p 'python3.buildEnv.override { extraLibs = [ python3Packages.jupyterlab_server ]; }'` look forward to this pull request!

File: pkgs/development/python-modules/jsonschema3/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{ stdenv, buildPythonPackage, fetchPypi, python
```
Comment (tbenst) [inline pkgs/development/python-modules/jsonschema3/default.nix]: This file is no longer needed as jsonschema has been updated to 3.1 already

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -2495,9 +2496,9 @@ in {
 
   jupyterlab_launcher = callPackage ../development/python-modules/jupyterlab_launcher { };
 
-  jupyterlab_server = callPackage ../development/python-modules/jupyterlab_server { };
+  jupyterlab_server = throw "pkgs.python3.pkgs.jupyterlab_server has moved to pkgs.jupyterlab_server";
```
Comment (tbenst) [inline pkgs/top-level/python-packages.nix]: I think we want to keep as a python module for reasons explained in https://github.com/NixOS/nixpkgs/pull/67423#issuecomment-538420909

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -3113,6 +3114,8 @@ in {
 
   jsonschema = callPackage ../development/python-modules/jsonschema { };
 
+  jsonschema3 = callPackage ../development/python-modules/jsonschema3 { };
```
Comment (tbenst) [inline pkgs/top-level/python-packages.nix]: no longer needed

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -4117,6 +4117,16 @@ in
 
   jupyter-kernel = callPackage ../applications/editors/jupyter/kernel.nix { };
 
+  inherit (
+    let pythonPkgs = python3.pkgs.overrideScope' (self: super: {
```
Comment (tbenst) [inline pkgs/top-level/all-packages.nix]: This is no longer necessary now that nixpkgs is on jsonschema-3.1

**Review by tbenst — COMMENTED** [review]
I think this can be simplified greatly now that we have jsonschema-3.1 in nixpkgs. I would suggest retaining this as a python-module, and getting rid of the overlays & various changes to top-level & python-modules/packages that are no longer necessary. Thanks for the work updating this!

**Comment by cdepillabout** [comment]
At work we are developing an extension for JupyterLab.  I'd really like to be able to write a nix expression to be able to create a development environment for JupyterLab extensions.

We are developing against a post-1.0 version of JupyterLab, so ideally I'd like to use this PR along with [`jupyterWith`](https://github.com/tweag/jupyterWith).

@Infinisil Is there anything you need help with to get this PR in a state where it can be merged in?

**Comment by evanjs** [comment]
Likewise wondering if there's anything that can be done to help.
Would love to swap Jupyter out for Jupyterlab (mmm, native dark theme)

**Comment by FRidh** [comment]
jupyterlab works fine with e3bcd8e33e2ab7fb63cbddc3d43da48a9506e89d

**Pull request closed without merging**